File: BinarySearchTree.js
class Node {
    constructor(value) {
        this.left = null;
        this.right = null;
        this.value = value;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }

    insert(value) {
        //  Check if value is smaller or greater than node,then go down right and left respectively.
        //   if is empty, assign there, if not, continue the same processing
        if (this.root.value == null) {
            this.root.value = value
            return;
        }
        let curr = this.root
        while (true) {
            if (value < curr.value) {
                if (curr.left) {
                    curr = curr.left
                } else {
                    curr.left = new Node(value)
                    return;
                }
            } else {
                if (curr.right) {
                    curr = curr.right
                } else {
                    curr.right = new Node(value)
                    return;
                }
            }


        }
    }

    lookup(value) {
        //Code here
    }

    // remove
}

const tree = new BinarySearchTree();
tree.insert(9)
tree.insert(4)
tree.insert(6)
tree.insert(20)
tree.insert(170)
tree.insert(15)
tree.insert(1)
JSON.stringify(traverse(tree.root))

//     9
//  4     20
//1  6  15  170

function traverse(node) {
    const tree = {value: node.value};
    tree.left = node.left === null ? null : traverse(node.left);
    tree.right = node.right === null ? null : traverse(node.right);
    return tree;
}

File: dataStractureAndAlgorithmes/binary_search_tree.py
class Node:
    def __init__(self, value):
        self.left = None
        self.right = None
        self.value = value


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        if self.root is None:
            self.root = Node(value)
            return

        curr = self.root
        while True:
            if value < curr.value:
                if curr.left:
                    curr = curr.left
                else:
                    curr.left = Node(value)
                    return
            else:
                if curr.right:
                    curr = curr.right
                else:
                    curr.right = Node(value)
                    return

    def lookup(self, value):
        curr = self.root
        while curr:
            if value == curr.value:
                return curr
            if value < curr.value:
                curr = curr.left
            else:
                curr = curr.right
        return None


tree = BinarySearchTree()
tree.insert(9)
tree.insert(4)
tree.insert(6)
tree.insert(20)
tree.insert(170)
tree.insert(15)
tree.insert(1)


def print_tree_indent(node, level=0):
    if node is None:
        return
    print_tree_indent(node.right, level + 1)
    print(' ' * 4 * level + '->', node.value)
    print_tree_indent(node.left, level + 1)


print_tree_indent(tree.root)
# print_tree_indent(tree.lookup(170))
# print_tree_indent(tree.lookup(122))

# print_tree_indent(tree.root)
